docs/oppen_algorithm.md

```diff
@@ -0,0 +1,179 @@
+```
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+| tokens_queue | BEGIN#1 | STRING#1 | BREAK | STRING#2 | BEGIN#2 | LINEBREAK | STRING#3 |
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+| size_buffer  |    ∞    |    5     |  -6   |    10    |   -18   |    -18    |    9     |
++--------------+---------+----------+-------+----------+---------+-----------+----------+
+
++-------------+-------+-------+-------+-------+
+| scan_stack  |   0   |   2   |   4   |   5   |
++-------------+-------+-------+-------+-------+
+
+Push BEGIN#1 to the print_stack.
+
++--------------+----------+-------+----------+---------+-----------+----------+
+| tokens_queue | STRING#1 | BREAK | STRING#2 | BEGIN#2 | LINEBREAK | STRING#3 |
++--------------+----------+-------+----------+---------+-----------+----------+
+| size_buffer  |    5     |  -6   |    10    |   -18   |    -18    |    9     |
++--------------+----------+-------+----------+---------+-----------+----------+
+
++-------------+-------+-------+-------+
+| scan_stack  |   2   |   4   |   5   |
++-------------+-------+-------+-------+
+
+Push STRING#1 to the print_stack.
+
++--------------+-------+----------+---------+-----------+----------+
+| tokens_queue | BREAK | STRING#2 | BEGIN#2 | LINEBREAK | STRING#3 |
++--------------+-------+----------+---------+-----------+----------+
+| size_buffer  |  -6   |    10    |   -18   |    -18    |    9     |
++--------------+-------+----------+---------+-----------+----------+
+
++-------------+-------+-------+-------+
+| scan_stack  |   2   |   4   |   5   |
++-------------+-------+-------+-------+
+```
+
+That process will keep on going until we flush the previous line and display `STRING#3` on a new line.
+
+```
++--------------+
+| tokens_queue |
++--------------+
+| size_buffer  |
++--------------+
+
++-------------+
+| scan_stack  |
++-------------+
+```
+
+Since `scan_stack` is empty, all the remaining tokens will directly be forwarded to the `print_stack` without having to pass by the other data structures.
+
+```
+Push STRING#4 to the print_stack.
+Push END#1    to the print_stack.
+Push END#2    to the print_stack.
+Push EOF      to the print_stack.
+```
+
+## Our extensions
+
+### Eager printing
+
+We extended Oppen's algorithm to allow groups to be printed eagerly: we will try to flush the current line every time an `END` token is encountered.
+
+### Indent anchors
+
+When nested groups are encountered, Oppen's algorithm anchors the next new line on the end of the previously displayed line. We extended the original algorithm in order to allow the anchor to be the current offset.
+
+### Trim trailing whitespaces
+
+Oppen's original work does not handle trailing whitespaces. We extended the algorithm to trim these trailing whitespaces.
+
```

Comment:
```suggestion
### Trim trailing whitespaces

It may have not been a concern to Oppen, but in a modern context, this is a nice feature to have, especially when running outputs through diff.

```